src/main/java/org/cryptomator/fusecloudaccess/OpenFileUploader.java

```diff
@@ -1,59 +1,85 @@
 package org.cryptomator.fusecloudaccess;
 
+import org.cryptomator.cloudaccess.api.CloudPath;
 import org.cryptomator.cloudaccess.api.CloudProvider;
 import org.cryptomator.cloudaccess.api.ProgressListener;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicLong;
 
+/**
+ * Prepares and schedules upload of (possibly) changed files to the cloud.
+ * <p>
+ * It should be ensured, that the files which are currently processed are not modified by other thread or processes.
+ */
 class OpenFileUploader {
 
 	private static final Logger LOG = LoggerFactory.getLogger(OpenFileUploader.class);
 
 	private final CloudProvider provider;
 	private final ConcurrentMap<Long, CompletionStage<Void>> scheduledUploads = new ConcurrentHashMap<>();
 	private final AtomicLong idGenerator = new AtomicLong();
+	private final Path cacheDir;
 
-	OpenFileUploader(CloudProvider provider) {
+	OpenFileUploader(CloudProvider provider, Path cacheDir) {
 		this.provider = provider;
+		this.cacheDir = cacheDir;
 	}
 
+	/**
+	 * Schedules {@link OpenFile} to be uploaded to the set {@link CloudProvider} by first creating a temporary copy of
+	 * it and start the upload by reading from the copy.
+	 * <p>
+	 * This method relies on that the underlying files in the file system do not change during the complete copy.
+	 * Generally it is not considered harmful, if a file is during upload replaced by a new version, since this version is also scheduled for an upload.
+	 *
+	 * @param file OpenFile object with reference to a real file.
+	 * @return A void {@link CompletionStage} indicating when the upload operation is successful completed.
+	 */
 	public CompletionStage<Void> scheduleUpload(OpenFile file) {
 		if (!file.isDirty()) {
 			LOG.trace("Upload of {} skipped. Unmodified.", file.getPath());
 			return CompletableFuture.completedFuture(null);
 		}
 		try {
-			var in = file.asPersistableStream();
-			return scheduleUpload(file, in);
+			Path toUpload = cacheDir.resolve(file.getPath().getFileName().toString() + ".tmp");
+			Files.copy(file.asPersistableStream(), toUpload, StandardCopyOption.REPLACE_EXISTING);
+			file.setDirty(false);
+			return scheduleUpload(file.getPath(), Files.newInputStream(toUpload, StandardOpenOption.DELETE_ON_CLOSE));
```

Comment:
I like the idea of `DELETE_ON_CLOSE`, but I'm not sure this works as intended on all OS.

Reply:
Doesn't mean this should be changed, but needs to be tested ;-)

CC @tobihagemann